sig-scalability/slos/pod_startup_latency.md

```diff
@@ -4,29 +4,43 @@
 
 | Status | SLI | SLO |
 | --- | --- | --- |
-| __Official__ | Startup latency of stateless<sup>[1](#footnote1)</sup> and schedulable<sup>[2](#footnote2)</sup> pods, excluding time to pull images and run init containers, measured from pod creation timestamp to when all its containers are reported as started and observed via watch, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, 99th percentile per cluster-day <= 5s |
+| __Official__ | Startup latency of schedulable<sup>[1](#footnote1)</sup> stateless<sup>[2](#footnote2)</sup> pods, excluding time to pull images and run init containers, measured from pod creation timestamp to when all its containers are reported as started and observed via watch, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, 99th percentile per cluster-day <= 5s |
+| __WIP__ | Startup latency of schedulable<sup>[1](#footnote1)</sup> stateful<sup>[3](#footnote3)</sup> pods, excluding time to pull images, run init containers and provision volumes (in delayed binding mode), measured from pod creation timestamp to when all its containers are reported as started and observed via watch, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, 99th percentile per cluster-day <= X where X depends on storage provider |
 
-<a name="footnote1">[1\]</a>A `stateless pod` is defined as a pod that doesn't
+<a name="footnote1">[1\]</a>By schedulable pod we mean a pod that can be
+scheduled in the cluster without causing any preemption.
+
+<a name="footnote2">[2\]</a>A `stateless pod` is defined as a pod that doesn't
 mount volumes with sources other than secrets, config maps, downward API and
 empty dir.
 
-<a name="footnote2">[2\]</a>By schedulable pod we mean a pod that can be
-scheduled in the cluster without causing any preemption.
+<a name="footnote3">[2\]</a>A `stateful pod` is defined as a pod that mounts
+at least one volume with sources other than secrets, config maps, downward API
+and empty dir.
 
 ### User stories
 - As a user of vanilla Kubernetes, I want some guarantee how quickly my pods
 will be started.
 
 ### Other notes
-- Only schedulable and stateless pods contribute to the SLI:
+- Only schedulable pods contribute to the SLIs:
   - If there is no space in the cluster to place the pod, there is not much
     we can do about it (it is task for Cluster Autoscaler which should have
     separate SLIs/SLOs).
   - If placing a pod requires preempting other pods, that may heavily depend
     on the application (e.g. on their graceful termination period). We don't
     want that to contribute to this SLI.
-  - Mounting disks required by non-stateless pods may potentially also require
-    non-negligible time, not fully dependent on Kubernetes.
+- We are explicitly splitting stateless and stateful pods from each other:
+  - Starting a stateful pod requires attaching and mounting volumes, that
```

Comment:
Another point to consider and I'm not sure if we want to explicitly include or exclude: not only does a stateful pod involve attaching and mounting volumes; it may also require waiting for the (same) volume used by a prior pod to be detached from the previous node.  So the startup of a stateful pod may also depend on waiting for the previous pod to be terminated/cleaned up.

Reply:
Great point.
For me, conceptually the situation is quite similar as for pods that need to preempt others (which is explicitly excluded) - basically we need to wait for something to happen (some pod is preempted or in this case a volume is unmounted).
So personally, I would say we should exclude it, but let me think a bit more about that (or let me know if you have any other thoughts about it).

Reply:
I think it can be excluded for now.  For a complete picture, having a pod deletion SLI should be enough to cover this scenario.  What I'm not sure about is in practice, how would we be able to automatically detect and exclude things like needing preemption or volume unmounting?

Reply:
I explicitly added unmounting volume part to the SLI.
Regarding detecting - it's great point - I should have been cleaner about it. We already kind-of have this problem (with schedulable) - it's easy to do that in tests (because we know what is happening there) and this is the first step. I added a TODO below to figure out how to do that in prod.